pkg/cmd/waitforkube/waitforkube.go

```diff
@@ -34,31 +37,53 @@ func NewWaitForKubeCommand(errOut io.Writer) *cobra.Command {
 				if err := fn(); err != nil {
 					if cmd.HasParent() {
 						klog.Fatal(err)
-						fmt.Fprint(waitForKubeOpts.errOut, err.Error())
+						fmt.Fprint(waitForKube.errOut, err.Error())
 					}
 				}
 			}
-			must(waitForKubeOpts.Run)
+			must(waitForKube.Run)
 		},
 	}
 
 	return cmd
 }
 
-func (w *waitForKubeOpts) Run() error {
+func (w *waitForKube) Run() error {
+	dataFilePath := filepath.Join(dataDir, dbPath)
+	if CheckEtcdDataFileExists(dataFilePath) {
```

Comment:
How will this affect the CEO use case of recovering single-member corrupted data directory?

Ideally the workflow is: crashloop detected by the operator -> member pushed to Remove state -> staticpodcontroller will remove the data directory -> operator will scale down -> the pod will restart -> CheckEtcdDataFileExists returns false, and scale up happens successfully

Is there a possibility of hitting CheckEtcdDataFileExists before operator detects crashloop or before the staticpodcontroller blows up the directory? if so we risk hitting a possible race condition.

Regardless we should create a test against openshift/origin to check single-member recover on ever PR, will create the corresponding issue!

Reply:
So workflow should be as you noted.

* `etcd-member` or `etcd-metrics` in crashloop backoff is observed by observation controller and status us set to MemberRemove[1]

* staticpodcontroller then stops the static pod and removes the directory[2].

> Is there a possibility of hitting CheckEtcdDataFileExists before operator detects crashloop or before the staticpodcontroller blows up the directory? if so we risk hitting a possible race condition.

If for some reason we hit `CheckEtcdDataFileExists` before we check for crashloop then the result is still valid. A peer exists until its data-dir is destroyed. we continue to etcd-member container and it crashloops controller removes after observation.

I think we can better handle the open areas of this potential race once we land https://github.com/openshift/cluster-etcd-operator/issues/24 as we can defer the status of the member more accurately.

So, for now, I believe this is needed for a few circumstances described https://github.com/openshift/cluster-etcd-operator/pull/73#issue-369909426

> Regardless we should create a test against openshift/origin to check single-member recover on ever PR, will create the corresponding issue!

I think this could fall into the disruptive test.

[1] https://github.com/openshift/cluster-etcd-operator/blob/6dcaeec687252d9bebe6cf97f6b31ca2a1273f66/pkg/operator/configobservation/etcd/observe_etcd.go#L449

[2] https://github.com/openshift/cluster-etcd-operator/blob/6dcaeec687252d9bebe6cf97f6b31ca2a1273f66/pkg/cmd/staticpodcontroller/staticpodcontroller.go#L233

[3] https://github.com/openshift/cluster-etcd-operator/blob/master/pkg/cmd/staticpodcontroller/staticpodcontroller.go#L221

Reply:
> How will this affect the CEO use case of recovering single-member corrupted data directory?

I will write up the various failure cases in UML and how they would be handled to make this more visually understandable.

Reply:
> If for some reason we hit CheckEtcdDataFileExists before we check for crashloop then the result is still valid. A peer exists until its data-dir is destroyed. we continue to etcd-member container and it crashloops controller removes after observation.

Makes sense, thanks for the explanation.